samza-api/src/main/java/org/apache/samza/metrics/RetainLastNPolicy.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.metrics;
+
+import java.util.Collection;
+import java.util.Iterator;
+
+
+public class RetainLastNPolicy<T> implements ListGaugeEvictionPolicy<T> {
+
+  private final ListGauge<T> listGauge;
+  private final int nItems;
+
+  public RetainLastNPolicy(ListGauge<T> listGauge, int numItems) {
+    this.listGauge = listGauge;
+    this.nItems = numItems;
+  }
+
+  @Override
+  public void elementAddedCallback() {
+    // get a snapshot of the list
+    Collection<T> listGaugeCollection = this.listGauge.getValue();
+    int numToEvict = listGaugeCollection.size() - nItems;
+    Iterator<T> iterator = listGaugeCollection.iterator();
+    while (numToEvict > 0 && iterator.hasNext()) {
+      // Remove in FIFO order to retain the last nItems
+      listGauge.remove(iterator.next());
```

Comment:
1. yes, an easy way is to bake in the eviction logic and expose only control-knows #items or durationThreshold, and remove policy as configurable, and get rid of remove altogether.

2. unmodifiableList  doesnt give a copy of the elements but it does give a copy of the references to the elements, and listGauge.remove is sync. 
so for most immutable elements (for which this was intended), this could work

Reply:
1. Above you mentioned that you would like to keep remove, but here it sounds like you are suggesting to get rid of remove. Could you please clarify?
2. I wasn't referring to multithreaded concurrency in this case. My concern was that the iterator you are iterating over in the while loop and the listGauge that you are calling remove on seem to be both backed by the same underlying list object. I was wondering if the removal can cause some inconsistency in the iterator. It could be helpful to have a removeHead() here too so you don't need to actually iterate through the elements.